### PR TITLE
Bugfix: Booking Success Redirect should be in parent frame in case of embed.

### LINKED ISSUE #8099: Redirection error
Redirection for embeds is now done directly at the embed location and not on a new page.

Bildschirmfoto 2023-04-05 um 12 51 17

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

Fixes #8099
Regression from #7947

**Environment**: Production

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

- [x] Create a Team and Individual Event and test that redirection happens with query params properly sent
- [x] Embed the page in an iframe and then see that redirect happens outside the iframe. Page containing the iframe gets redirected
- [x] Test a paid event and verify that after payment redirection happens to custom success page.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 1]
```diff
@@ -264,15 +264,31 @@ const BookingPage = ({
         );
       }
 
+      const query = {
```
Comment (hariombalhara) [inline apps/web/components/booking/pages/BookingPage.tsx:267]: Moved query object here.

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 2]
```diff
@@ -264,15 +264,31 @@ const BookingPage = ({
...
+        isSuccessBookingPage: true,
+        email: bookingForm.getValues("responses.email"),
+        eventTypeSlug: eventType.slug,
+        seatReferenceUid: "seatReferenceUid" in responseData ? responseData.seatReferenceUid : null,
+        ...(rescheduleUid && booking?.startTime && { formerTime: booking.startTime.toString() }),
+      };
+
+      if (eventType.successRedirectUrl) {
+        const url = new URL(eventType.successRedirectUrl);
+        Object.entries(query).forEach(([key, value]) => {
```
Comment (hariombalhara) [inline apps/web/components/booking/pages/BookingPage.tsx:277]: Add search params related to booking to existing redirect URL that might have it's own query params

File: apps/web/pages/booking/[uid].tsx [hunk 3]
```diff
@@ -1013,15 +1013,6 @@ export async function getServerSideProps(context: GetServerSidePropsContext) {
     customInputs: customInputSchema.array().parse(eventTypeRaw.customInputs),
   };
 
-  if (eventType.successRedirectUrl && isSuccessBookingPage && !cancel) {
```
Comment (hariombalhara) [inline apps/web/pages/booking/[uid].tsx:1016]: Don't redirect here as server isn't aware of whether the page is embedded or not.

File: apps/web/components/booking/pages/BookingPage.tsx [hunk 4]
```diff
@@ -264,32 +265,38 @@ const BookingPage = ({
...
-        },
+      const query = {
+        isSuccessBookingPage: true,
+        email: bookingForm.getValues("responses.email"),
+        eventTypeSlug: eventType.slug,
+        seatReferenceUid: "seatReferenceUid" in responseData ? responseData.seatReferenceUid : null,
+        ...(rescheduleUid && booking?.startTime && { formerTime: booking.startTime.toString() }),
+      };
+
+      return bookingSuccessRedirect({
```
Comment (hariombalhara) [inline apps/web/components/booking/pages/BookingPage.tsx:276]: Do redirection client side which can do redirect outside the iframe. Server can't do that.

File: packages/lib/bookingSuccessRedirect.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    const url = new URL(successRedirectUrl);
+    Object.entries(query).forEach(([key, value]) => {
+      if (value === null || value === undefined) {
+        return;
+      }
+      url.searchParams.append(key, String(value));
+    });
+
+    // Using parent ensures, Embed iframe would redirect outside of the iframe.
+    window.parent.location.href = url.toString();
```
Comment (hariombalhara) [inline packages/lib/bookingSuccessRedirect.ts:26]: When the page isn't in an iframe. `window.parent === window`.

File: packages/lib/bookingSuccessRedirect.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  query: Record<string, string | null | undefined | boolean>;
+  router: NextRouter;
+}) => {
+  if (successRedirectUrl) {
+    const url = new URL(successRedirectUrl);
+    Object.entries(query).forEach(([key, value]) => {
+      if (value === null || value === undefined) {
+        return;
+      }
+      url.searchParams.append(key, String(value));
```
Comment (hariombalhara) [inline packages/lib/bookingSuccessRedirect.ts:22]: Append query params to possibly already present params in success url

**Review by Udit-takkar — APPROVED** [review]
https://user-images.githubusercontent.com/53316345/230970615-f738f80a-424e-40c5-8394-391cda727760.mov

LGTM

**Pull request merged**